pandas/tests/frame/test_dtypes.py

```diff
@@ -815,6 +815,19 @@ def test_astype_extension_dtypes_duplicate_col(self, dtype):
         expected = concat([a1.astype(dtype), a2.astype(dtype)], axis=1)
         tm.assert_frame_equal(result, expected)
 
+    def test_df_where_with_category(self):
+        # GH 16979
+        df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
+        mask = np.array([[True, False, True], [False, True, True]])
+        # change type to category
+        df.A = df.A.astype("category")
+        df.B = df.B.astype("category")
+        df.C = df.C.astype("category")
+
+        expected = df.A.where(mask[:, 0])
+        result = df.A.where(mask[:, 0], other=None)
+        tm.assert_series_equal(result, expected)
+
```

Comment:
Similar to my comments in https://github.com/pandas-dev/pandas/pull/29455, let's parameterize as follows:

~~~python
@pytest.mark.parametrize("kwargs", [dict(), dict(other=None)])
def test_df_where_with_category(self, kwargs):
    df = DataFrame(np.arange(2 * 3).reshape(2, 3), columns=list("ABC"))
    mask = np.array([[True, False, True], [False, True, True]])

    # Change type to category
    df.A = df.A.astype("category")
    df.B = df.B.astype("category")
    df.C = df.C.astype("category")

    expected = ...   # Construct expected result
    result = df.A.where(mask[:, 0], **kwargs)

    tm.assert_series_equal(result, expected)
~~~